File: jdbc/src/main/scala/org/beangle/tomcat/jdbc/EncryptedDataSourceFactory.scala
package org.beangle.tomcat.jdbc

import java.io.BufferedReader
import java.io.InputStreamReader
import java.{ util => ju }
import java.net.{ HttpURLConnection, URL }

import javax.script.ScriptEngineManager
import javax.naming.Context
import javax.sql.DataSource

import org.apache.tomcat.jdbc.pool.DataSourceFactory
import org.apache.tomcat.jdbc.pool.PoolConfiguration
import org.apache.tomcat.jdbc.pool.XADataSource
import scala.language.existentials
/**
 * Encrypted DataSourceFactory
 *
 * @author chaostone
 * @see http://www.idesign4all.nl/blog/?p=103
 */
class EncryptedDataSourceFactory extends DataSourceFactory {

  override def createDataSource(properties: ju.Properties, context: Context, XA: Boolean): DataSource = {
    val url = properties.get("url").asInstanceOf[String]
    if (null != url && url.startsWith("http")) properties.putAll(parse(getResponseText(new URL(url))))

    val poolProperties = DataSourceFactory.parsePoolProperties(properties)
    val encodedPwd = poolProperties.getPassword
    val keyName = poolProperties.getName.replace("/", "_") + "_secret"
    val secretKey = System.getenv(keyName)
    val password =
      if (encodedPwd.startsWith("?")) {
        if (null == secretKey) decryptByPrompt(keyName, encodedPwd.substring(1))
        else new Encryptor(secretKey).decrypt(encodedPwd.substring(1))
      } else {
        new Encryptor(secretKey).decrypt(encodedPwd)
      }
    poolProperties.setPassword(password)

    if (poolProperties.getDataSourceJNDI != null && poolProperties.getDataSource == null) performJNDILookup(context, poolProperties)
    val dataSource = if (XA) new XADataSource(poolProperties) else new org.apache.tomcat.jdbc.pool.DataSource(poolProperties)
    dataSource.createPool()
    dataSource
  }

  /**
   * decrypt password using user input.
   */
  private def decryptByPrompt(keyName: String, encodedPwd: String): String = {
    var i = 0
    var correctSecret = false
    var decoded: String = null
    val br = new BufferedReader(new InputStreamReader(System.in))
    Console.print(s"What is the $keyName:")
    while (i < 3 && !correctSecret) {
      var secretKey: String = null
      try {
        secretKey = br.readLine()
        if (null == secretKey) {
          Console.print(s"What is the $keyName key:")
        } else {
          decoded = new Encryptor(secretKey).decrypt(encodedPwd)
          correctSecret = true
        }
      } catch {
        case e: Throwable => Console.print("Incorrect key(" + secretKey + "),try again:")
      }
      i += 1
    }
    if (null == decoded) throw new RuntimeException(s"Cannot decoded $encodedPwd")
    decoded
  }

  private def parse(string: String): ju.Properties = {
    val sem = new ScriptEngineManager
    val engine = sem.getEngineByName("javascript")
    val result = new ju.Properties
    val iter = engine.eval("result =" + string).asInstanceOf[ju.Map[_, _]].entrySet().iterator()
    while (iter.hasNext) {
      val one = iter.next.asInstanceOf[ju.Map.Entry[_, AnyRef]]
      val value = one.getValue match {
        case d: java.lang.Double =>
          if (java.lang.Double.compare(d, d.intValue) > 0) d.toString
          else String.valueOf(d.intValue)
        case a: Any => a.toString
      }
      result.put(one.getKey.toString, value)
    }
    println(result)
    result
  }

  private def getResponseText(constructedUrl: URL): String = {
    var conn: HttpURLConnection = null
    try {
      conn = constructedUrl.openConnection().asInstanceOf[HttpURLConnection]
      var in: BufferedReader = null
      in = new BufferedReader(new InputStreamReader(conn.getInputStream, "UTF-8"))
      var line: String = in.readLine()
      val stringBuffer = new StringBuffer(255)
      stringBuffer.synchronized {
        while (line != null) {
          stringBuffer.append(line)
          stringBuffer.append("\n")
          line = in.readLine()
        }
        stringBuffer.toString
      }
    } catch {
      case e: Exception => throw new RuntimeException(e)
    } finally {
      if (conn != null) conn.disconnect()
    }
  }
}

File: core/src/main/scala/org/beangle/sas/tomcat/EncryptedDataSourceFactory.java
/*
 * Beangle, Agile Development Scaffold and Toolkit
 *
 * Copyright (c) 2005-2017, Beangle Software.
 *
 * Beangle is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Beangle is distributed in the hope that it will be useful.
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Beangle.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.beangle.sas.tomcat;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Iterator;
import java.util.Map;
import java.util.Properties;

import javax.naming.Context;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.sql.DataSource;

import org.apache.tomcat.jdbc.pool.DataSourceFactory;
import org.apache.tomcat.jdbc.pool.PoolConfiguration;
import org.apache.tomcat.jdbc.pool.XADataSource;

/**
 * Encrypted DataSourceFactory
 *
 * @author chaostone
 * @see http://www.jdev.it/encrypting-passwords-in-tomcat/
 */
class EncryptedDataSourceFactory extends DataSourceFactory {

  @Override
  public DataSource createDataSource(Properties properties, Context context, boolean XA) throws Exception {
    String url = (String) properties.get("url");
    if (null != url && url.startsWith("http")) properties.putAll(parse(getResponseText(new URL(url))));

    PoolConfiguration poolProperties = DataSourceFactory.parsePoolProperties(properties);
    String encodedPwd = poolProperties.getPassword();
    String keyName = poolProperties.getName().replace("/", "_") + "_secret";
    String secretKey = System.getenv(keyName);
    String password = null;

    if (encodedPwd.startsWith("?")) {
      if (null == secretKey) password = decryptByPrompt(keyName, encodedPwd.substring(1));
      else password = new Encryptor(secretKey).decrypt(encodedPwd.substring(1));
    } else {
      password = new Encryptor(secretKey).decrypt(encodedPwd);
    }
    poolProperties.setPassword(password);

    if (poolProperties.getDataSourceJNDI() != null && poolProperties.getDataSource() == null) {
      performJNDILookup(context, poolProperties);
    }
    if (XA) {
      XADataSource ds = new XADataSource(poolProperties);
      ds.createPool();
      return ds;
    } else {
      org.apache.tomcat.jdbc.pool.DataSource ds = new org.apache.tomcat.jdbc.pool.DataSource(poolProperties);
      ds.createPool();
      return ds;
    }
  }

  /**
   * decrypt password using user input.
   */
  private String decryptByPrompt(String keyName, String encodedPwd) {
    int i = 0;
    boolean correctSecret = false;
    String decoded = null;
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    System.out.print("What is the " + keyName + ":");
    while (i < 3 && !correctSecret) {
      String secretKey = null;
      try {
        secretKey = br.readLine();
        if (null == secretKey) {
          System.out.print("What is the " + keyName + " key:");
        } else {
          decoded = new Encryptor(secretKey).decrypt(encodedPwd);
          correctSecret = true;
        }
      } catch (Throwable e) {
        System.out.print("Incorrect key(" + secretKey + "),try again:");
      }
      i += 1;
    }
    if (null == decoded) throw new RuntimeException("Cannot decoded " + encodedPwd);
    return decoded;
  }

  private Properties parse(String string) throws Exception {
    ScriptEngineManager sem = new ScriptEngineManager();
    ScriptEngine engine = sem.getEngineByName("javascript");
    Properties result = new Properties();
    @SuppressWarnings({ "unchecked", "rawtypes" })
    Iterator<Map.Entry<?, Object>> iter = ((Map) (engine.eval("result =" + string))).entrySet().iterator();
    while (iter.hasNext()) {
      Map.Entry<?, Object> one = iter.next();
      String value = null;
      if (one.getValue() instanceof Double) {
        Double d = (Double) one.getValue();
        if (java.lang.Double.compare(d, d.intValue()) > 0) value = d.toString();
        else value = String.valueOf(d.intValue());
      } else {
        value = one.getValue().toString();
      }

      String key = (one.getKey().toString() == "maxActive") ? "maxTotal" : one.getKey().toString();
      result.put(key, value);
    }
    return result;
  }

  private String getResponseText(URL constructedUrl) {
    HttpURLConnection conn = null;
    try {
      conn = (HttpURLConnection) constructedUrl.openConnection();
      BufferedReader in = new BufferedReader(new InputStreamReader(conn.getInputStream(), "UTF-8"));
      String line = in.readLine();
      StringBuffer stringBuffer = new StringBuffer(255);
      while (line != null) {
        stringBuffer.append(line);
        stringBuffer.append("\n");
        line = in.readLine();
      }
      return stringBuffer.toString();
    } catch (Exception e) {
      throw new RuntimeException(e);
    } finally {
      if (conn != null) conn.disconnect();
    }
  }
}
